clarity/src/vm/ast/parser_v2/lexer/mod.rs

```diff
@@ -0,0 +1,1957 @@
+                        self.proceed_through_error(LexerError::InvalidCharIdent(self.next));
+                        return ident;
+                    }
+                }
+            }
+            self.read_char();
+        }
+    }
+
+    fn read_trait_identifier(&mut self) -> String {
+        let start_line = self.last_line as u32;
+        let start_column = self.last_column as u32;
+        let mut ident = String::new();
+
+        loop {
+            match self.next {
+                'a'..='z' | 'A'..='Z' | '0'..='9' | '_' | '-' => ident.push(self.next),
+                '>' => return ident,
+                _ => {
+                    if is_separator(self.next) {
+                        self.add_diagnostic(PlacedError {
+                            span: Span {
+                                start_line: self.line as u32,
+                                start_column: self.column as u32,
+                                end_line: self.line as u32,
+                                end_column: self.column as u32,
+                            },
+                            e: LexerError::ExpectedClosing('>'),
+                        });
+                        self.add_diagnostic(PlacedError {
+                            span: Span {
+                                start_line,
+                                start_column,
+                                end_line: start_line,
+                                end_column: start_column,
+                            },
+                            e: LexerError::NoteToMatchThis('<'),
+                        });
+                        return ident;
+                    } else {
+                        self.proceed_through_error(LexerError::InvalidCharTraitIdent(self.next));
+                        return ident;
+                    }
+                }
+            }
+            self.read_char();
+        }
+    }
+
+    fn read_principal(&mut self) -> String {
+        let mut principal = String::new();
+
+        loop {
+            self.read_char();
+            match self.next {
+                'A'..='Z' | '0'..='9' => principal.push(self.next),
+                _ => {
+                    if is_separator(self.next) {
+                        return principal;
+                    } else {
+                        self.proceed_through_error(LexerError::InvalidCharPrincipal(self.next));
+                        return principal;
+                    }
+                }
+            }
+        }
+    }
+
+    fn read_unsigned(&mut self) -> u128 {
+        let mut num: u128 = 0;
+        while self.next.is_ascii_digit() {
+            let digit = self.next as u32 - '0' as u32;
+            num = num * 10 + digit as u128;
+            self.read_char();
+        }
+        if !is_separator(self.next) {
+            self.proceed_through_error(LexerError::InvalidCharUint(self.next));
+        }
+        num
+    }
+
+    fn read_integer(&mut self) -> i128 {
```

Comment:
I have the same concern here as with `read_unsigned`.  What happens if there is a literal that requires more than 128 bits to represent?

Reply:
Fixed in a43454bd2

Reply:
Rebased, now 639e4030e348158f63c0b8fbacbb52083b95fb74